### PR TITLE
ApplyHyperspace rule - JoinIndexRule

### PR DESCRIPTION
**Pull request by sezruby**
### What is the context for this pull request?

 - **Tracking Issue**: n/a
 - **Parent Issue**: #405 
 - **Dependencies**: n/a

### What changes were proposed in this pull request?

Refactored `JoinIndexRule` based on `IndexFilter`

`ApplyHyperspace` will be substitute `JoinIndexRule` & `FilterIndexRule`  when it's delivered completely.

### Does this PR introduce _any_ user-facing change?

No

### How was this patch tested?

Unit test

### CODE REVIEW

File: src/main/scala/com/microsoft/hyperspace/index/rules/disabled/JoinIndexRule.scala [hunk 1]
```diff
@@ -0,0 +1,661 @@
...
+import com.microsoft.hyperspace.index.rankers.JoinIndexRanker
+import com.microsoft.hyperspace.index.rules._
+import com.microsoft.hyperspace.index.rules.ApplyHyperspace.{PlanToIndexesMap, PlanToSelectedIndexMap}
+import com.microsoft.hyperspace.index.rules.disabled.JoinAttributeFilter.extractConditions
+import com.microsoft.hyperspace.index.sources.FileBasedRelation
+import com.microsoft.hyperspace.shim.JoinWithoutHint
+import com.microsoft.hyperspace.telemetry.{AppInfo, HyperspaceEventLogging, HyperspaceIndexUsageEvent}
+import com.microsoft.hyperspace.util.ResolverUtils.resolve
+
+object JoinPlanNodeFilter extends QueryPlanIndexFilter {
```
Comment (sezruby) [inline src/main/scala/com/microsoft/hyperspace/index/rules/disabled/JoinIndexRule.scala:37]: from JoinIndexRule.isApplicable, except for ensureAttributeRequirements

File: src/main/scala/com/microsoft/hyperspace/index/rules/disabled/JoinIndexRule.scala [hunk 2]
```diff
@@ -0,0 +1,661 @@
...
+    if (indexes.isEmpty || indexes.size != 2) {
+      return Map.empty
+    }
+
+    if (withReasonTagAll(
+          "Each join condition column should come from " +
+            "relations directly and attributes from left plan must exclusively have " +
+            "one-to-one mapping with attributes from right plan. " +
+            "E.g. join(A = B and A = D) is not eligible.") {
+          ensureAttributeRequirements(
```
Comment (sezruby) [inline src/main/scala/com/microsoft/hyperspace/index/rules/disabled/JoinIndexRule.scala:150]: from JoinIndexRule.ensureAttributeRequirement

File: src/main/scala/com/microsoft/hyperspace/index/rules/disabled/JoinIndexRule.scala [hunk 3]
```diff
@@ -0,0 +1,661 @@
...
+    condition match {
+      case EqualTo(_: AttributeReference, _: AttributeReference) =>
+        Seq(condition)
+      case And(left, right) =>
+        extractConditions(left) ++ extractConditions(right)
+      case _ => throw new IllegalStateException("Unsupported condition found")
+    }
+}
+
+object JoinColumnFilter extends QueryPlanIndexFilter {
```
Comment (sezruby) [inline src/main/scala/com/microsoft/hyperspace/index/rules/disabled/JoinIndexRule.scala:282]: from JoinIndexRule.getBestIndexPair (except for JoinIndexRanker)

File: src/main/scala/com/microsoft/hyperspace/index/rules/disabled/JoinIndexRule.scala [hunk 4]
```diff
@@ -0,0 +1,661 @@
...
+object JoinRankFilter extends IndexRankFilter {
+  override def apply(plan: LogicalPlan, indexes: PlanToIndexesMap): PlanToSelectedIndexMap = {
+    if (indexes.isEmpty || indexes.size != 2) {
+      return Map.empty
+    }
+
+    val leftRelation = JoinIndexRule.leftRelation.get
+    val rightRelation = JoinIndexRule.rightRelation.get
+    val lRMap = JoinIndexRule.leftToRightColumnMap.get
+    val compatibleIndexPairs =
```
Comment (sezruby) [inline src/main/scala/com/microsoft/hyperspace/index/rules/disabled/JoinIndexRule.scala:473]: from JoinIndexRule.getBestIndexPair, getCompatibleIndexPairs & JoinIndexRanker

File: src/main/scala/com/microsoft/hyperspace/index/rules/IndexFilter.scala [hunk 5]
```diff
@@ -62,6 +62,39 @@ trait IndexFilter extends ActiveSparkSession {
...
+   * Append the reason string to WHYNOT_REASON tag for the given list of indexes
+   * if the result of the function is false and WHYNOT_ENABLED tag is set to the index.
+   *
+   * @param reasonString Informational message in case condition is false.
+   * @param f Function for a condition
+   * @param keyPair A pair of plan and index list to tag
+   * @return Result of the given function
+   */
+  protected def withReasonTagAll(reasonString: => String)(f: => Boolean)(
+      implicit keyPair: (LogicalPlan, Seq[IndexLogEntry])): Boolean = {
```
Comment (sezruby) [inline src/main/scala/com/microsoft/hyperspace/index/rules/IndexFilter.scala:76]: Created a new function with a different name as overloading on implicit parameter doesn't work

**Review by clee704 — APPROVED** [review]
LGTM, thanks!

But it would've been easier if refactoring changes were a separate PR.

**Comment by sezruby** [comment]
Let me merge this change as it is for quick delivery of refactoring.
Feel free to make any comment in this PR.

**Pull request merged**